lib/rack/lock.rb

```diff
@@ -12,15 +12,15 @@ def initialize(app, mutex = Mutex.new)
     end
 
     def call(env)
-      old, env[FLAG] = env[FLAG], false
       @mutex.lock
-      response = @app.call(env)
-      body = BodyProxy.new(response[2]) { @mutex.unlock }
-      response[2] = body
-      response
-    ensure
-      @mutex.unlock unless body
-      env[FLAG] = old
+      begin
+        response = @app.call(env.merge(FLAG => false))
+        body = BodyProxy.new(response[2]) { @mutex.unlock }
```

Comment:
I think we can use `Array#fetch` here instead `Array#[]` as we know the index `2` will exist. And if not we want to know it also via the exception.